superduperdb/components/graph.py

```diff
@@ -0,0 +1,267 @@
+import dataclasses as dc
+import typing as t
+from collections import OrderedDict
+from functools import cached_property
+from inspect import signature
+
+import networkx as nx
+
+from superduperdb.base.artifact import Artifact
+from superduperdb.components.component import Component
+from superduperdb.components.model import Model
+from superduperdb.ext.torch.model import TorchModel
+
+if t.TYPE_CHECKING:
+    from superduperdb.base.datalayer import Datalayer
+
+
+class KeyStore:
+    def __init__(self, model, params):
+        self.model = model
+        self._params = params
+        for param in params:
+            setattr(self, param, f'{model}.{param}')
+
+    def __len__(self):
+        return len(self._params)
+
+    @property
+    def params(self):
+        return self._params
+
+
+class RootModel:
+    def __init__(self):
+        self.input = KeyStore('root', [])
+
+    def predict(self, x):
+        return x
+
+
+class GraphModel:
+    def __init__(self, model: t.Union['Model', 'TorchModel']):
+        self.identifier = model.identifier
+        self.output = f'{model.identifier}.output'
+
+        predict_method = model.predict_method
+        if isinstance(model, TorchModel):
+            predict_method = model.forward_method
+
+        assert isinstance(model.object, Artifact)
+        self.object = model.object.artifact
+
+        if predict_method:
+            self.predict_method = getattr(self.object, predict_method)
+        else:
+            self.predict_method = self.object
+
+        params = self.set_input()
+        self.input = KeyStore(model.identifier, params)
+
+    def set_input(self):
+        sig = signature(self.predict_method)
+        return list(sig.parameters.keys())
+
+    def predict(self, *args, **kwargs):
+        return self.predict_method(*args, **kwargs)
+
+
+class Node:
```

Comment:
Do we need this? Can't we simply use `Model` instances as the nodes.

Reply:
Yes, this is needed this holds states for Model which we should not do it in Model

Reply:
You can keep this information directly in the Graph

Reply:
This is creating too much unnecessary boilerplate.